src/main/java/com/uber/cadence/internal/replay/ReplayDecider.java

```diff
@@ -382,13 +386,40 @@ private void handleWorkflowExecutionSignaled(HistoryEvent event) {
   public DecisionResult decide(PollForDecisionTaskResponse decisionTask) throws Throwable {
     lock.lock();
     try {
-      boolean forceCreateNewDecisionTask = decideImpl(decisionTask, null);
-      return new DecisionResult(decisionsHelper.getDecisions(), forceCreateNewDecisionTask);
+      AtomicReference<Map<String, WorkflowQueryResult>> queryResults = new AtomicReference<>();
+      boolean forceCreateNewDecisionTask =
+          decideImpl(
+              decisionTask, () -> queryResults.set(getQueryResults(decisionTask.getQueries())));
+      return new DecisionResult(
+          decisionsHelper.getDecisions(), queryResults.get(), forceCreateNewDecisionTask);
     } finally {
       lock.unlock();
     }
   }
 
+  private Map<String, WorkflowQueryResult> getQueryResults(Map<String, WorkflowQuery> queries) {
+    if (queries == null) {
+      return null;
+    }
+
+    return queries
+        .entrySet()
+        .stream()
+        .collect(Collectors.toMap(q -> q.getKey(), q -> queryWorkflow(q.getValue())));
+  }
+
+  private WorkflowQueryResult queryWorkflow(WorkflowQuery query) {
```

Comment:
Does this happen after the history from the current decision has already been applied? It will violate correctness if we first run queries and then apply the history in the decision.

Reply:
This happens after all decision logic is executed. This can be seen from 392 line in this file - a callback to get query results is provided into `decideImpl` which is called in the `finally` clause (507 line) after all decision logic.